packages/analytics-js/src/app/RudderAnalytics.ts

```diff
@@ -133,6 +141,74 @@ class RudderAnalytics implements IRudderAnalytics<IAnalytics> {
     this.getAnalyticsInstance(writeKey).load(writeKey, dataPlaneUrl, loadOptions);
   }
 
+  /**
+   * A function to track page lifecycle events like page loaded and page unloaded
+   * @param preloadedEventsArray
+   * @param loadOptions
+   * @returns
+   */
+  // eslint-disable-next-line class-methods-use-this
+  trackPageLifecycleEvents(
+    preloadedEventsArray: PreloadedEventCall[],
+    loadOptions?: Partial<LoadOptions>,
+  ) {
+    const { trackPageLifecycle, useBeacon } = loadOptions ?? {};
+    const { events = [], enabled = false, options = {} } = trackPageLifecycle ?? {};
+    if (enabled) {
+      const visitId = generateUUID();
+      const pageLoadedTimestamp = Date.now();
+      if (events.length === 0 || events.includes(PageLifecycleEvents.PAGELOADED)) {
+        preloadedEventsArray.unshift([
+          'track',
+          PageLifecycleEvents.PAGELOADED,
+          { visitId },
+          {
+            originalTimestamp: new Date(pageLoadedTimestamp).toISOString(),
+            ...options,
+          },
+        ]);
+      }
+      if (events.length === 0 || events.includes(PageLifecycleEvents.PAGEUNLOADED)) {
+        if (useBeacon === true) {
+          // Register the page unloaded lifecycle event listeners
+          onPageLeave((isAccessible: boolean) => {
+            if (isAccessible === false) {
+              const visitDuration = Date.now() - pageLoadedTimestamp;
```

Comment:
We should store the current timestamp here again and override it as the original timestamp through the options as before.
Even if the user calculates the duration again from the timestamps in the events, it should perfectly match the `visitDuration` field.

Reply:
Addressed